sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/SupportsDynamicFiltering.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.read;
+
+import org.apache.spark.annotation.Experimental;
+import org.apache.spark.sql.connector.expressions.NamedReference;
+import org.apache.spark.sql.sources.Filter;
+
+/**
+ * A mix-in interface for {@link Scan}. Data sources can implement this interface if they can
+ * dynamically filter {@link InputPartition}s that were originally planned using predicates
+ * Spark infers by reusing parts of the query.
+ *
+ * @since 3.2.0
+ */
+@Experimental
+public interface SupportsDynamicFiltering extends Scan {
+  /**
+   * Returns attributes this scan can be dynamically filtered by.
+   * <p>
+   * Spark will call {@link #filter(Filter[])} if it can derive a dynamic
+   * predicate for any of the filter attributes.
+   */
+  NamedReference[] filterAttributes();
+
+  /**
+   * Dynamically filters this scan.
+   * <p>
+   * The provided expressions must be interpreted as a set of filters that are ANDed together.
+   * Implementations may use the filters to prune originally planned {@link InputPartition}s.
+   * <p>
+   * Spark will call {@link Scan#toBatch()}, {@link Scan#toMicroBatchStream(String)} or
+   * {@link Scan#toContinuousStream(String)} again after filtering the scan dynamically.
+   * The newly produced {@link Batch} or its streaming alternative may report a subset of
+   * originally planned {@link InputPartition}s.
+   *
+   * @param filters dynamic filters
+   */
+  void filter(Filter[] filters);
```

Comment:
The design doc has alternative ways to represent dynamic filters. It would be great to get feedback on this.

Reply:
ds v2 API prefers to use the native data classes in Spark, e.g. InternalRow, UTF8String, etc. However, we keep using the v1 `Filter` API which uses external data classes. Shall we consider adding a v2 Filter API which uses v2 `Expression` and native data classes?

Reply:
I think that would be best, @cloud-fan. Has there been any discussion on how the new API should look like? Since the old API has been exposed in `SupportsPushDownFilters`, what is the plan for introducing the new API? Will we introduce a new method with a default implementation that would translate into to the old API?

Reply:
@dbtsai used to have a PR on this: https://github.com/dbtsai/spark/pull/10 but hasn't been updated for a while now.

Reply:
Thanks for the pointer, @sunchao. I think @rdblue's comment [here](https://github.com/dbtsai/spark/pull/10/files#r368072666) matches my proposal above about adding v2 filters in parallel and having a default implementation that converts v2 to v1. 

I can pick up the v2 filter API but I'd like to do that independently of this PR if everyone is on board.

Reply:
We can probably have a `SupportsPushDownV2Filters` to not break existing code.

Reply:
I'll be fine with either adding `SupportsPushDownV2Filters` or just adding new methods with a default implementation.

```
interface SupportsPushDownFilters extends ScanBuilder {
  Filter[] pushFilters(Filter[] filters);

  default V2Filter[] pushFilters(V2Filter[] filters) {
    // convert v2 to v1 and call pushFilters for v1
    // convert v1 result back to v2
  }
}
```

In any case, we must ensure we don't break existing sources.

Reply:
Is it alright with everybody to consider v1 filters in the scope of this PR? I'll take over @dbtsai's PR later.

Reply:
If we target this PR for 3.3, then I'm fine to use v1 Filter here and replace it with v2 Filter later, as there is plenty of time. Otherwise, I'd like to have v2 Filter first, to avoid releasing this API with v1 Filter and breaking it later.

Reply:
I think that this should use v1 filters. The new filters don't exist yet and the DSv2 API uses v1 in other places. There is no need to block this on adding v2 filters. If the release for this is the same for v2 filters, we can consider removing support. But at this point I think we should not assume that will happen.

Reply:
Since dynamic filtering can provide substantial performance improvements for v2 tables, I'd love to get this feature into 3.2. As we already use v1 filters in other Data Source V2 interfaces, I feel it should be alright to use them here too.

As I noted above, we don't really have to break this API once we have v2 filters. We can follow whatever we decide to do with `SupportsPushDownFilters`: introduce a separate interface or just add a method to the existing interface with a default implementation that would convert v2 filters into v1 filters.

Does this seem reasonable, @cloud-fan?

Reply:
> If we target this PR for 3.3, then I'm fine to use v1 Filter here and replace it with v2 Filter later, as there is plenty of time. Otherwise, I'd like to have v2 Filter first, to avoid releasing this API with v1 Filter and breaking it later.

Do we plan to remove v1 Filter soon? Otherwise, we still can keep v1 Filter support in this API even we decide to add v2 Filter later. So it seems that we don't actually break it quickly (at least not in next release).